### PR TITLE
Implement socks5 proxy support using socksio

### PR DESCRIPTION
**Pull request by cdeler**
The final draft PR for https://github.com/encode/httpx/issues/203 (I closed previous). It's based on https://github.com/encode/httpcore/pull/51, I added @yeraydiazdiaz to co-authored-by

What I did:
1. added `pproxy` as a dev. dependency to have a local socks proxy server
1. added `AsyncSOCKSConnection` and `SyncSOCKSConnection`. For now it handles only socks5-without-auth connections

### CODE REVIEW

File: httpcore/_async/connection.py [hunk 1]
```diff
@@ -162,3 +164,76 @@ async def aclose(self) -> None:
         async with self.request_lock:
             if self.connection is not None:
                 await self.connection.aclose()
+
+
+class AsyncSOCKSConnection(AsyncHTTPConnection):
```
Comment (florimondmanca) [inline httpcore/_async/connection.py:169]: A bit wary of the subclassing dependency here...

I understand the intent that "a SOCKS connection is just an HTTP connection, but connected differently". But since "open a socket" seems to be the *only* step that changes, and considering we have a `socket` parameter on `AsyncHTTPConnection`, would it make sense for the "open a SOCKS proxy socket" operation to be performed by an `AsyncSOCKSProxy` class?

In other words, instead of a separate connection, we'd have a new proxy connection pool class that opens a socket explicitly before instantiating and saving a connection.

Dunno how feasible that is and I haven't went through actually implementing this, and perhaps you've already thought about that, but just throwing ideas out there! :)

**Review by florimondmanca — COMMENTED** [review]
Nice start! Curious if CI passes with this setup :eyes:

File: httpcore/_async/connection.py (hunk 1 above)
Comment (cdeler) [reply httpcore/_async/connection.py:169]: You mean to replace inheritance by aggregation?

Comment (florimondmanca) [reply httpcore/_async/connection.py:171]: Sure, never mind, I see that #51 took this approach too. :-)

**Comment by cdeler** [comment]
Well, I mostly implemented `socks5` proxy support. Now I want to think up what to do with `socks4`

**Comment by cdeler** [comment]
Converted back to draft until socks4 proxy support has been added (then we can split the PR into chunks)

**Comment by cdeler** [comment]
I reworked this draft PR according to the @tomchristie 's comment: https://gitter.im/encode/community?at=5f6db8edf41f4105e4d7a9eb

So that right now `(A)syncSocksProxy`  is the wrapper over the `(A)syncConnectionPool`

The next step I want to do is to merge https://github.com/encode/httpcore/pull/199 (it's a PR about migrating `http` proxy suite from `mitmproxy` to `pproxy`)

**Comment by tomchristie** [comment]
Now supported in 0.14.5 - https://www.encode.io/httpcore/proxies/#socks-proxy-support
